sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/AmqpEventHubConsumer.cs

```diff
@@ -130,11 +161,103 @@ public override void UpdateRetryPolicy(EventHubRetryPolicy newRetryPolicy)
         ///
         /// <returns>The batch of <see cref="EventData" /> from the Event Hub partition this consumer is associated with.  If no events are present, an empty enumerable is returned.</returns>
         ///
-        public override Task<IEnumerable<EventData>> ReceiveAsync(int maximumMessageCount,
-                                                                  TimeSpan? maximumWaitTime,
-                                                                  CancellationToken cancellationToken)
+        public override async Task<IEnumerable<EventData>> ReceiveAsync(int maximumMessageCount,
+                                                                        TimeSpan? maximumWaitTime,
+                                                                        CancellationToken cancellationToken)
         {
-             throw new NotImplementedException();
+            Argument.AssertNotClosed(_closed, nameof(AmqpEventHubClient));
+            Argument.AssertAtLeast(maximumMessageCount, 1, nameof(maximumMessageCount));
+
+            var receivedEventCount = 0;
+            var failedAttemptCount = 0;
+            var waitTime = (maximumWaitTime ?? _tryTimeout);
+            var retryDelay = default(TimeSpan?);
+            var amqpMessages = default(IEnumerable<AmqpMessage>);
+            var receivedEvents = default(List<EventData>);
+
+            var stopWatch = Stopwatch.StartNew();
+
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    try
+                    {
+                        EventHubsEventSource.Log.EventReceiveStart(EventHubName, ConsumerGroup, PartitionId);
+
+                        ReceivingAmqpLink link = await ReceiveLink.GetOrCreateAsync(UseMinimum(ConnectionScope.SessionTimeout, _tryTimeout)).ConfigureAwait(false);
+                        cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+                        var messagesReceived = await Task.Factory.FromAsync
+                        (
+                            (callback, state) => link.BeginReceiveMessages(maximumMessageCount, waitTime, callback, state),
+                            (asyncResult) => link.EndReceiveMessages(asyncResult, out amqpMessages),
+                            TaskCreationOptions.RunContinuationsAsynchronously
+                        ).ConfigureAwait(false);
+
+                        cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+
+                        // If event messages were received, then package them for consumption and
+                        // return them.
+
+                        if ((messagesReceived) && (amqpMessages != null))
+                        {
+                            receivedEvents ??= new List<EventData>();
```

Comment:
That's a new operator! (ie. ??=) Need to read about this.

Reply:
Yup, it's a shortcut to that.  :)